// @elastic/react-search-ui-views/es/layouts/Layout.js
import PropTypes from "prop-types";
import React from "react";
import LayoutSidebar from "./LayoutSidebar";
import { appendClassName } from "../view-helpers";

function Layout(_ref) {
  var className = _ref.className,
      children = _ref.children,
      header = _ref.header,
      bodyContent = _ref.bodyContent,
      bodyFooter = _ref.bodyFooter,
      bodyHeader = _ref.bodyHeader,
      sideContent = _ref.sideContent;
  return React.createElement("div", {
    className: appendClassName("sui-layout", className)
  }, React.createElement("div", {
    className: "sui-layout-header"
  }, React.createElement("div", {
    className: "sui-layout-header__inner"
  }, header)), React.createElement("div", {
    className: "sui-layout-body"
  }, React.createElement("div", {
    className: "sui-layout-body__inner"
  }, React.createElement(LayoutSidebar, {
    className: "sui-layout-sidebar"
  }, sideContent), React.createElement("div", {
    className: "sui-layout-main"
  }, React.createElement("div", {
    className: "sui-layout-main-header"
  }, React.createElement("div", {
    className: "sui-layout-main-header__inner"
  }, bodyHeader)), React.createElement("div", {
    className: "sui-layout-main-body"
  }, children || bodyContent), React.createElement("div", {
    className: "sui-layout-main-footer"
  }, bodyFooter)))));
}

Layout.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  header: PropTypes.node,
  bodyContent: PropTypes.node,
  bodyFooter: PropTypes.node,
  bodyHeader: PropTypes.node,
  sideContent: PropTypes.node
};
export default Layout;
