// @elastic/react-search-ui-views/es/layouts/LayoutSidebar.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import PropTypes from "prop-types";
import React from "react";
import { appendClassName } from "../view-helpers";

var LayoutSidebar =
/*#__PURE__*/
function (_React$Component) {
  _inherits(LayoutSidebar, _React$Component);

  function LayoutSidebar(props) {
    var _this;

    _classCallCheck(this, LayoutSidebar);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(LayoutSidebar).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "toggleSidebar", function () {
      _this.setState(function (_ref) {
        var isSidebarToggled = _ref.isSidebarToggled;
        return {
          isSidebarToggled: !isSidebarToggled
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "renderToggleButton", function (label) {
      if (!_this.props.children) return null;
      return React.createElement("button", {
        hidden: true,
        type: "button",
        className: "sui-layout-sidebar-toggle",
        onClick: _this.toggleSidebar
      }, label);
    });

    _this.state = {
      isSidebarToggled: false
    };
    return _this;
  }

  _createClass(LayoutSidebar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children;
      var isSidebarToggled = this.state.isSidebarToggled;
      var classes = appendClassName(className, isSidebarToggled ? "".concat(className, "--toggled") : null);
      return React.createElement(React.Fragment, null, this.renderToggleButton("Show Filters"), React.createElement("div", {
        className: classes
      }, this.renderToggleButton("Save Filters"), children));
    }
  }]);

  return LayoutSidebar;
}(React.Component);

_defineProperty(LayoutSidebar, "propTypes", {
  className: PropTypes.string,
  children: PropTypes.node
});

export default LayoutSidebar;
