// @elastic/react-search-ui-views/lib/layouts/LayoutSidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _viewHelpers = require("../view-helpers");

var LayoutSidebar =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(LayoutSidebar, _React$Component);

  function LayoutSidebar(props) {
    var _this;

    (0, _classCallCheck2.default)(this, LayoutSidebar);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(LayoutSidebar).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "toggleSidebar", function () {
      _this.setState(function (_ref) {
        var isSidebarToggled = _ref.isSidebarToggled;
        return {
          isSidebarToggled: !isSidebarToggled
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderToggleButton", function (label) {
      if (!_this.props.children) return null;
      return _react.default.createElement("button", {
        hidden: true,
        type: "button",
        className: "sui-layout-sidebar-toggle",
        onClick: _this.toggleSidebar
      }, label);
    });
    _this.state = {
      isSidebarToggled: false
    };
    return _this;
  }

  (0, _createClass2.default)(LayoutSidebar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children;
      var isSidebarToggled = this.state.isSidebarToggled;
      var classes = (0, _viewHelpers.appendClassName)(className, isSidebarToggled ? "".concat(className, "--toggled") : null);
      return _react.default.createElement(_react.default.Fragment, null, this.renderToggleButton("Show Filters"), _react.default.createElement("div", {
        className: classes
      }, this.renderToggleButton("Save Filters"), children));
    }
  }]);
  return LayoutSidebar;
}(_react.default.Component);

(0, _defineProperty2.default)(LayoutSidebar, "propTypes", {
  className: _propTypes.default.string,
  children: _propTypes.default.node
});
var _default = LayoutSidebar;
exports.default = _default;
